service/history/shard/context_impl.go

```diff
@@ -1805,80 +1794,69 @@ func (s *ContextImpl) acquireShard() {
 	// 3. The state changes to Stopping or Stopped.
 	//
 	// If the shard controller sees that service resolver has assigned ownership to someone
-	// else, it will call finishStop, which will trigger case 3 above.
+	// else, it will call finishStop, which will trigger case 3 above, and also cancel
+	// lifecycleCtx. The persistence operations called here use lifecycleCtx as their context,
+	// so if we were blocked in any of them, they should return immediately with a context
+	// canceled error.
 	policy := backoff.NewExponentialRetryPolicy(1 * time.Second)
 	policy.SetExpirationInterval(5 * time.Minute)
 
 	// Remember this value across attempts
 	ownershipChanged := false
 
 	op := func() error {
+		if !s.isValid() {
+			return ErrShardClosed
+		}
+
 		// Initial load of shard metadata
 		err := s.loadShardMetadata(&ownershipChanged)
 		if err != nil {
 			return err
 		}
 
-		s.wLock()
-		defer s.wUnlock()
-
-		// Check that we should still be running
-		if s.state >= contextStateStopping {
-			return errStoppingContext
-		}
-
 		// Try to acquire RangeID lock. If this gets a persistence error, it may call:
-		// transitionLocked(contextRequestStop) for ShardOwnershipLostError:
-		//   This will transition to Stopping right here, and the transitionLocked call at the end of the
+		// transition(contextRequestStop) for ShardOwnershipLostError:
+		//   This will transition to Stopping right here, and the transition call at the end of the
 		//   outer function will do nothing, since the state was already changed.
-		// transitionLocked(contextRequestLost) for other transient errors:
+		// transition(contextRequestLost) for other transient errors:
 		//   This will do nothing, since state is already Acquiring.
+		s.wLock()
 		err = s.renewRangeLocked(true)
 		if err != nil {
+			s.wUnlock()
```

Comment:
why not wUnlock right after call to renewRangeLocked()?

Reply:
oh duh, I'll fix that